Generate `flink-conf.yaml` file automatically to set optimum conf values

chandrashekar-s opened the pull request:
## Description of what I changed

Fixes #823 
- Enabled a feature flag to automate the generation of `flink-conf.yaml` based on the `numThreads` in `application.yaml` file and the number of `cores` available in the machine.
- If the flag is disabled, then the `flink-conf.yaml` directed by the env. var `FLINK_CONF_DIR` will be used. If var is empty, then default conf will be used.
- JAVA_OPTS can be set to override the default JVM heap and other parameter values.
-  The application will fail to launch if the JVM Off Heap memory is insufficient for the Flink cluster to launch.
- Fixes the regression introduced by this [change](https://github.com/google/fhir-data-pipes/pull/670) where the maxWorkers was defaulted to 1 and in turn the num of shards for parquet files was also 1.

## E2E test

Tested `Full Run` and `Incremental Run` by enabling and disabling the flags. Also tested for different `numThread` configurations.

TESTED:

Load tested the application on a 48 core machine. Given below are the results

Run # | Run Type | # of Patient Records | Time in secs (Before Change) | Time in secs (After Change) | Remarks
-- | -- | -- | -- | -- | --
1 | Full | 9K | 400 | 358 | Full run timing remains the same before and after the changes
2 | Full (Repeat) | 9K | 360 | 256 |  
3 | Incremental | 9K Upfront + 100 additional | 140 | 30 | The improvement in time after change is because of fixing the regression (increased num of shards from 1 to N)
4 | Incremental | 9K Upfront + 500 additional | 180 | 32 |  

After fixing the regression the time taken for incremental run has significantly reduced

## Checklist: I completed these to help reviewers :)

<!--- Put an `x` in the box if you did the task -->
<!--- If you forgot a task please follow the instructions below -->

- [x] I have read and will follow the [review process](https://github.com/GoogleCloudPlatform/openmrs-fhir-analytics/blob/master/doc/review_process.md).
- [x] I am familiar with Google Style Guides for the language I have coded in.

  No? Please take some time and review [Java](https://google.github.io/styleguide/javaguide.html) and [Python](https://google.github.io/styleguide/pyguide.html) style guides.

- [x] My IDE is configured to follow the Google [**code styles**](https://google.github.io/styleguide/).

  No? Unsure? -> [configure your IDE](https://github.com/google/google-java-format).

- [x] I have **added tests** to cover my changes. (If you refactored existing code that was well tested you do not have to add tests)
- [x] I ran `mvn clean package` right before creating this pull request and added all formatting changes to my commit.
- [x] All new and existing **tests passed**.
- [x] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

chandrashekar-s commented:
@bashir2 The build is failing currently, I am looking into it.

chandrashekar-s commented:
@bashir2 The PR is ready for review. Since changes were made to create target jars compatible with jdk 17, the docker images had to be updated for streaming module.

bashir2 commented on pipelines/controller/src/test/java/com/google/fhir/analytics/FlinkConfigurationTest.java:
```diff
@@ -0,0 +1,72 @@
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+
+public class FlinkConfigurationTest {
```
nit: Please use Mickito i.e., `@RunWith(MockitoJUnitRunner.class)` both for its benefits (e.g., easier mocking) and consistency.

chandrashekar-s replied:
Done

bashir2 commented on pipelines/controller/src/test/java/com/google/fhir/analytics/FlinkConfigurationTest.java:
```diff
@@ -0,0 +1,72 @@
+
+public class FlinkConfigurationTest {
+
+  @ParameterizedTest
```
Nice, I did not know this feature, thanks :-)

bashir2 commented on pipelines/controller/src/test/java/com/google/fhir/analytics/FlinkConfigurationTest.java:
```diff
@@ -0,0 +1,72 @@
+  private static Stream<Arguments> provideFlinkParams() {
+    return Stream.of(
+        // The percentage completion is rounded off to next highest integer
+        Arguments.of(2, getNetworkMemory(2)), Arguments.of(2, getNetworkMemory(2)));
```
Why are we repeating the same argument twice?

chandrashekar-s replied:
Sorry my bad. Updated to have the different values in the latest commit.

bashir2 commented on pipelines/controller/src/test/java/com/google/fhir/analytics/FlinkConfigurationTest.java:
```diff
@@ -0,0 +1,72 @@
+  @MethodSource("provideFlinkParams")
+  public void testConfigurations(int numThreads, long expectedNetworkMemory) throws IOException {
+    DataProperties dataProperties = new DataProperties();
+    dataProperties.setAutoGenerateFlinkConfiguration(true);
```
Can we test the `false` scenario too?

chandrashekar-s replied:
Done

bashir2 reviewed (commented):
Thanks @chandrashekar-s for this change.

chandrashekar-s commented on pipelines/batch/src/main/java/com/google/fhir/analytics/ParquetMerger.java:
```diff
@@ -64,31 +64,50 @@ public class ParquetMerger {
+                    out.output(KV.of(id, record));
+                  }
+                }))
+        .apply(GroupByKey.create());
```
As part of this change, we now read all parquet files for a given resource type at once. This has the benefit of reducing the number of Flink operations by one. Have added the comments in the code.

chandrashekar-s commented on pipelines/controller/pom.xml:
```diff
@@ -201,6 +201,30 @@
+                <manifestEntries>
+                  <!-- These options enable access to all methods in the mentioned packages at
+                  compile and runtime. These options are added as MANIFEST entries in the bundled
+                  jar  -->
```
Done

chandrashekar-s commented on pipelines/controller/src/main/java/com/google/fhir/analytics/FlinkConfiguration.java:
```diff
@@ -0,0 +1,167 @@
+      Path confPath = Files.createTempDirectory(TEMP_FLINK_CONF_DIR);
+      logger.info("Creating Flink temporary configuration directory at {}", confPath);
+      Path filePath = Paths.get(confPath.toString(), GlobalConfiguration.FLINK_CONF_FILENAME);
+      validateAndCreateConfFile(filePath, dataProperties);
```
The configuration file that gets created in Temp folder or the one that gets set by the env. variable FLINK_CONF_DIR gets passed to the Flink Execution Environment as per this [method](https://github.com/apache/beam/blob/444e10d29e04e3e144ef08c0a505bd76aa08984f/runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkExecutionEnvironments.java#L87).

However for non-local modes there are many configuration parameters that could be used, so I think we should recommend the users to pass their own custom file instead of Auto generated one.

bashir2 replied:
Can you please document it somewhere, i.e., in one of the README files or maybe one of the Wiki pages? I think it is fine for now as I have not seen a user with a Flink cluster (so Flink is usually used locally); but we should leave that option open (horizontal scalability).

chandrashekar-s replied:
Added checks in the code to invoke this code if it is local and autoGenerate is enabled. In case of non-local mode this will be by-passed. Documented in the `pipelines/controller/apllication.yaml` file as well

chandrashekar-s commented on pipelines/controller/src/main/java/com/google/fhir/analytics/PipelineManager.java:
```diff
@@ -382,12 +405,20 @@ synchronized void runIncrementalPipeline()
+        dataProperties.getNumThreads() > 0
+            ? dataProperties.getNumThreads()
+            : Runtime.getRuntime().availableProcessors();
+    mergerOptions.setNumShards(numShards);
```
Yes, this fixes the regression introduced where after the regression the number of shards (number of parquet files) was defaulted to one for incremental runs. Updated the PR description.

bashir2 replied:
Here too it seems to me this is mostly works for local runs. If it is a cluster run, we should change this; but we can investigate this as part of the issue that you file for 4 reshuffle in ParquetIO (maybe mention this there too).

chandrashekar-s replied:
Filed an [issue](https://github.com/google/fhir-data-pipes/issues/893) to handle this scenario for cluster mode.

chandrashekar-s commented on pipelines/controller/src/main/java/com/google/fhir/analytics/PipelineManager.java:
```diff
@@ -382,12 +405,20 @@ synchronized void runIncrementalPipeline()
+    flinkOptionsForMerge.setMaxParallelism(dataProperties.getMaxWorkers());
     if (dataProperties.getNumThreads() > 0) {
-      flinkOptions.setParallelism(dataProperties.getNumThreads());
+      flinkOptionsForMerge.setParallelism(dataProperties.getNumThreads());
```
Yes, this could be set to `numShards`, but I felt leaving it to Flink would be best. Internally Flink uses the similar logic to determine parallelism if the numThreads is not > 0. Setting this to `numShards` may be consistent so that the number of Shards and parallelism is always the same (Its the same even now as well).

bashir2 replied:
I see, so even when `numShards` is equal to the number of pipeline threads in ParquetIO, we still get 4 reshuffles, correct?

chandrashekar-s replied:
Yes, right. The number of reshuffle operations for writing to new parquet files remains the same (i.e 4), irrespective of the `numShards`.

chandrashekar-s commented on pipelines/controller/src/main/java/com/google/fhir/analytics/FlinkConfiguration.java:
```diff
@@ -0,0 +1,167 @@
+            "Flink configuration will be initialised from the directory at {}", envFlinkConfDir);
+        configuration = GlobalConfiguration.loadConfiguration(envFlinkConfDir);
+      }
+      configuration = TaskExecutorResourceUtils.adjustForLocalExecution(configuration);
```
These changes are valid for local mode. I think to make it generic (for both local and non-local mode) some more checks needs to be made.

bashir2 replied:
Okay closing this for now; I think adding documentation is good enough for now.

chandrashekar-s commented on pipelines/pom.xml:
```diff
@@ -304,7 +304,20 @@
+          <target>17</target>
+          <compilerArgs>
+            <!-- Enable access to the packages at compile time -->
+            <arg>--add-exports=java.base/jdk.internal.misc=ALL-UNNAMED</arg>
```
This change was needed at both the places i.e. at `pipelines/pom.xml` so that the packages are available during compile and run times (test cases), as well as `pipelines/controller/pom.xml` so that these options gets added as manifest entries in the `bundled` fat jar.

chandrashekar-s commented on pipelines/controller/src/main/java/com/google/fhir/analytics/FlinkConfiguration.java:
```diff
@@ -0,0 +1,167 @@
+
+  // This value is arrived based on the maximum number of reshuffle/partition operations that can
+  // occur in the pipelines. This has to be updated if any modifications are made to the pipelines
+  // that impact the reshuffle/partition operations.
```
Documentation has been made on how the number `8` is determined. Also, to avoid regressions, I not sure how the Junit test cases can be written. However, we can enable e2e tests for incremental run and validate if the autogenerated flink config runs fine for the pipelines.

chandrashekar-s replied:
Have added an item in this [issue](https://github.com/google/fhir-data-pipes/issues/891), for checking the number of reshuffle operations in a pipeline. Will be closing this comment for now.

chandrashekar-s commented on pipelines/batch/src/main/java/com/google/fhir/analytics/ParquetMerger.java:
```diff
@@ -64,31 +64,50 @@ public class ParquetMerger {
+        pipeline
+            .apply(Create.of(getParquetFilePaths(resourceType, dwhFilesList)))
+            .apply(FileIO.matchAll())
+            .apply(FileIO.readMatches());
```
I have verified the number of the parallel worker threads that will be used after this change and it remains the same as before. I am assuming performance tests should not be part of the CI as it will delay the build. I am open for suggestions on this.

bashir2 replied:
We can have a non-blocking performance test, i.e., it won't block a merge but we run it on each commit such that if a performance regression is introduced we can pinpoint the culprit quickly; but that's for #607. For this change if you can run the pipeline on a larger input, e.g., the 8K Patient set (with 1.6M Observations) and compare before/after run-time, that would be great.

chandrashekar-s replied:
Benchmarking is in progress, will update the results once done.

chandrashekar-s replied:
I have attached the benchmark results in the PR description. The performance seems to have improved after fixing the regression.

chandrashekar-s commented:
@bashir2 Thanks for reviewing the changes. I have addressed/responded to the comments. Can you please have a look. Also, for few cases where unit test cases could not be written, I feel we need few more tests for e2e or performance to avoid regressions.

bashir2 commented on pipelines/batch/src/main/java/com/google/fhir/analytics/ParquetMerger.java:
```diff
@@ -64,31 +64,56 @@ public class ParquetMerger {
+  private static PCollection<KV<String, Iterable<GenericRecord>>> readAndGroupById(
+      Pipeline pipeline, List<DwhFiles> dwhFilesList, String resourceType) {
+
+    // Reading all parquet files at once instead of one at a time, reduces the number of Flink
```
nit:
```suggestion
    // Reading all parquet files at once instead of one set at a time, reduces the number of Flink
```

chandrashekar-s replied:
Done

bashir2 commented on pipelines/controller/src/main/java/com/google/fhir/analytics/FlinkConfiguration.java:
```diff
@@ -41,10 +41,24 @@ public class FlinkConfiguration {
+   * <ul>
+   *   <li>2 reshuffle operations for matching and reading from Parquet files
+   *   <li>1 reshuffle operation for GroupByKey
+   *   <li>4 reshuffle operations for writing back to new parquet files (some kind of map-reduce
```
Can you please file a bug for this such that we investigate it in future? We should not really need 4 reshuffles there (it is ParquetIO that is doing this and maybe we can fix it or configure it to avoid this).

chandrashekar-s replied:
Have created this [issue](https://github.com/google/fhir-data-pipes/issues/891) for the same.

bashir2 commented on pipelines/controller/config/application.yaml:
```diff
@@ -105,6 +105,13 @@ fhirdata:
+  # have the parameters set with optimised values necessary to run the pipelines without fail.
+  # Disable this parameter to manually pass the configuration file by pointing the environment
+  # variable FLINK_CONF_DIR to the directory where the flink-conf.yaml is placed.
+  autoGenerateFlinkConfiguration: true
```
Here is also a good place to document that this option should be set to `false` if it is not a local run.

chandrashekar-s replied:
Done

bashir2 reviewed (approved):
Thanks @chandrashekar-s the remaining comments are all questions or minor suggestions. Please feel free to merge after addressing them.

bashir2 commented on pipelines/controller/src/main/java/com/google/fhir/analytics/FlinkConfiguration.java:
```diff
@@ -0,0 +1,181 @@
+        totalDirectMemory * EtlUtils.NO_OF_PARALLEL_PIPELINES;
+
+    long jvmMaxDirectMemory = jdk.internal.misc.VM.maxDirectMemory();
+    if (jvmMaxDirectMemory <= totalDirectMemoryOfAllParallelPipelines) {
```
So as discussed off-PR, let's make sure that the memory validation does not break the non-local runs, i.e., when the pipeline is running on a cluster.

chandrashekar-s commented:
Thanks @bashir2 for the reviewing the changes. I have addressed some of the comments in the latest commit. Also I have created these 2 issues #893 and #891 for `Validating Flink in non-local mode` and `Inestigating/fixing the reshuffle operations for writing to parquet files` respectively.

chandrashekar-s commented:
The performance results have been attached in the PR description. No noticeable changes for the `Full run`, but for `Incremental run` the timing has improved after fixing the number of Shards to be `N` (Flink parallelism)
